tests_e2e/tests/agent_status/agent_status.py

```diff
@@ -0,0 +1,143 @@
+        # Validate vm_agent_version field
+        vm_agent_version = instance_view.vm_agent.vm_agent_version
+        if vm_agent_version is None:
+            raise Exception("Instance view is missing agent version")
+        elif 'Unknown' in vm_agent_version:
+            raise Exception("Instance view shows agent version is unknown")
+
+        # Validate statuses field
+        statuses = instance_view.vm_agent.statuses
+        if statuses is None:
+            raise Exception("Instance view is missing agent statuses")
+        elif len(statuses) < 1:
+            raise Exception("Instance view is missing an agent status entry")
+        else:
+            self.validate_instance_view_vmagent_status(instance_view=instance_view)
+
+    def validate_instance_view(self, instance_view: VirtualMachineInstanceView):
+        """
+        Checks that instance view has vm_agent.statuses and vm_agent.vm_agent_version properties which report the Guest
+        Agent as running and Ready:
+
+        "vm_agent": {
+            "extension_handlers": [],
+            "vm_agent_version": "9.9.9.9",
+            "additional_properties": {},
+            "statuses": [
+                {
+                    "level": "Info",
+                    "time": "<class 'datetime.datetime'>",
+                    "message": "Guest Agent is running",
+                    "code": "ProvisioningState/succeeded",
+                    "additional_properties": {},
+                    "display_status": "Ready"
+                }
+            ]
+        }
+        """
+        if instance_view.vm_agent is None:
+            raise Exception("Instance view is missing vm agent")
+        else:
+            self.validate_instance_view_vmagent(instance_view=instance_view)
+
+        if instance_view.statuses is None:
+            raise Exception("Instance view is missing statuses")
+
+        log.info("Instance view has valid agent status, agent version: {0}, status: {1}"
+                 .format(instance_view.vm_agent.vm_agent_version, instance_view.vm_agent.statuses[0].display_status))
+
+    def run(self):
+        log.info("")
+        log.info("*******Verifying the agent status*******")
+
+        vm = VirtualMachineClient(self._context.vm)
+
+        timeout = datetime.datetime.now() + datetime.timedelta(minutes=5)
+        instance_view_is_valid = False
+        instance_view_exception = None
+
+        # Retry validating instance view with timeout of 5 minutes
+        while datetime.datetime.now() <= timeout and not instance_view_is_valid:
+            instance_view = vm.get_instance_view()
+            log.info("")
+            log.info(
+                "Check instance view to validate that the Guest Agent reports status without any new goal states...")
+            log.info("Instance view of VM is:\n%s", json.dumps(instance_view.serialize(), indent=2))
+
+            try:
+                self.validate_instance_view(instance_view)
```

Comment:
How are we validating that the Agent is reporting status with no goal state? This seems to just be checking that the instance view has an Agent status with the expected format, but that status could have been reported as the result of a previous goal state.

Reply:
I updated to check that the agent status timestamp in the instance view updates & we check that the last processed goal state in the agent log is the same

Reply:
Thanks